### PR TITLE
Triggers

### PR DESCRIPTION
**Pull request by ldanilek**
Implement Triggers. See the README for how to use it, and also an overview of how it works.

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/convex-helpers/server/triggers.ts [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    const triggers = this;
+    return {
+      args: {},
+      input(ctx: Ctx) {
+        const innerDb = ctx.db;
+        return { ctx: {
+          db: new DatabaseWriterWithTriggers(ctx, innerDb, triggers),
+          innerDb,
+        }, args: {} };
+      },
```
Comment (sujayakar) [inline packages/convex-helpers/server/triggers.ts:81]: you could bind `this` by doing...
```ts
input: (ctx: Ctx) => {
   ...
   new DatabaseWriterWithTriggers(ctx, innerDb, this),
},
```

File: packages/convex-helpers/server/triggers.ts [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+  async lock(): Promise<(() => void)> {
+    while (this.promise !== null) {
+      await this.promise;
+    }
+    [this.promise, this.resolve] = this._newLock();
+    return () => {
+      this.promise = null;
+      this.resolve?.();
+    };
+  }
```
Comment (sujayakar) [inline packages/convex-helpers/server/triggers.ts:99]: would it make sense to do a `withLock` API so it's much harder to forget to unlock?

File: packages/convex-helpers/server/triggers.ts [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+ *     other triggers recursively.
+ * - Recursive triggers are queued up, so they are executed in the same order
+ *   as the database writes were. At a high level, this is a BFS traversal of
+ *   the trigger graph.
+ * - Note when there are multiple triggers, they can't be executed atomically
+ *   with the writes that caused them, from the perspective of the other
+ *   triggers. So if one trigger is making sure denormalized data is
+ *   consistent, another trigger could see the data in an inconsistent state.
+ *   To avoid such problems, triggers should be resilient to such
+ *   inconsistencies or the trigger graph should be kept simple.
```
Comment (sujayakar) [inline packages/convex-helpers/server/triggers.ts:127]: do you think it'd be reasonable to just _ban_ recursive triggers for now? that way the semantics would be very straightforward. but perhaps that's too restrictive?

**Review by sujayakar — APPROVED** [review]
cool! some minor stuff inline but feel free to land as is.

File: packages/convex-helpers/server/triggers.ts (hunk 3 above)
Comment (ldanilek) [reply packages/convex-helpers/server/triggers.ts:127]: you actually get pretty similar guarantees when you have multiple triggers, even if they're not recursive.

like if you have these triggers:

```
triggers.register("users", async (ctx, change) => {
  const isTopStudent = await userAggregate.offsetUntil(ctx, change.newDoc.gpa) < 5;
  if (change.newDoc.isTopStudent !== isTopStudent) {
    ctx.db.patch(change.id, { isTopStudent });
  }
});
triggers.register("users", userAggregate.triggers());
```

one of the triggers has to account for the other not having run yet.

having multiple triggers seems like an important use-case, so if we have those then we might as well have recursive triggers too?

I think the semantics of recursive triggers makes sense, and I wouldn't look forward to banning them -- would it throw an error or just not execute them? -- and explaining those semantics.

**Pull request merged**